tests/st/calico_upgrade/test_calico_upgrade.py

```diff
@@ -0,0 +1,292 @@
+            logger.debug(
+                "INFO: calico-upgrade dry-run %s output:\n%s" % (report2, st_report2))
+            assert dr_report2 == st_report2, \
+                "INFO: calico-upgrade dry-run and start %s files are not equal" % report2
+
+    @parameterized.expand([
+        ("ippool_mixed", False),
+        ("policy_long_name", False),
+        ("profile_big", False),
+    ])
+    @attr('slow')
+    def test_start_abort_ignore_v3_data(self, testname, fail_expected):
+        """
+        Test the abort command re-enables the v1 Ready flag.
+
+        calico-upgrade abort - aborts the upgrade process
+        by resuming Calico networking for the v2.x nodes.
+        """
+        testdata = data[testname]
+
+        calicoctlv2("create", data=testdata)
+        logger.debug("INFO: dump of etcdv2:")
+        dump_etcdv2()
+
+        rcu = calicoupgrade("start")
+        logger.debug("INFO: calico-upgrade start should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        assert ready_output == "false"
+
+        datastore_ready_rc = _get_ready_etcdv3()
+        assert datastore_ready_rc is False
+
+        rcu = calicoupgrade("abort")
+        logger.debug("INFO: calico-upgrade abort should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        assert ready_output == "true"
+
+        datastore_ready_rc = _get_ready_etcdv3()
+        assert datastore_ready_rc is False
+
+        rcu = calicoupgrade("start --ignore-v3-data")
+        logger.debug("INFO: calico-upgrade start should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        assert ready_output == "false"
+
+        datastore_ready_rc = _get_ready_etcdv3()
+        assert datastore_ready_rc is False
+
+        rcu = calicoupgrade("complete")
+        logger.debug("INFO: calico-upgrade complete should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        assert ready_output == "false"
+
+        datastore_ready_rc = _get_ready_etcdv3()
+        assert datastore_ready_rc is True
+
+
+@parameterized.expand([
+    "dry-run",
+    "start",
+])
+def test_empty_datastore(cmd):
+    """
+    Test dry-run and start when the etcdv2 datastore is empty.
+    """
+    wipe_etcdv2(get_ip())
+    dump_etcdv2()
+    rcu = calicoupgrade(cmd)
+    logger.debug("INFO: empty etcdv2 datastore: Expecting a non-zero"
+                 + "return code for calico-upgrade %s.", cmd)
+    rcu.assert_error()
+
+
+def _get_readlines(file_generated):
+    """
+    Get the readlines() of the file generated.
+    """
+    file_output = open("/code/calico-upgrade-report/"+file_generated, 'rU')
+    file_output_lines = file_output.readlines()
+    file_output.close()
+    return file_output_lines
+
+
+def _get_ready_etcdv3():
```

Comment:
I guess this works ok, but you could have just used calicoctl v3 to query the cluster information resource rather than querying etcd directly.